internal/weaver/types.go

```diff
@@ -0,0 +1,31 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package weaver
+
+import "reflect"
+
+var (
+	// IsImplements, IsRef, and IsListener return whether the provided type is
+	// a weaver.Implements[T], weaver.Ref[T], and weaver.Listener respectively.
+	// These values are populated by an init function in the main weaver
+	// package. The circuitousness is to avoid cyclic dependencies.
+	//
+	// TODO(mwhittaker): Improve these abstractions to initialize values as
+	// well. The code in fillrefs, filllisteners, etc. relies on some unsafe
+	// reflection stuff.
+	IsImplements func(reflect.Type) bool
```

Comment:
You must have thought of it, but would it work if you defined `Implements` (and other structs) in this file, and weaver.go had:

```
type Implements[T any] weaver_internal.Implements[T]
type Ref[T any] weaver_internal.Ref[T]
...
```

Reply:
I like having the type directly in weaver.go. I suspect it is better in godoc for users of the system.

Reply:
Yeah, I had tried this but also preferred the godoc being simpler. I'll probably think harder about it because there might be a way to have nice godoc and simplify things.

Reply:
I looked into this more. I was able to replace `IsImplements` with something simpler, but `IsRef` and `IsListener`&mdash;and more generally, the way we initialize `weaver.Ref`s and `weaver.Listener`s&mdash;are hard to simplify even if all the code were in the same package. When we go to initialize a `weaver.Listener` field in a component struct, for example, the field may not be exported. Because of this, we're forced to use the icky unsafe parts of the reflect package. I'll open a PR with some slight simplifications though.